python/Ganga/GPIDev/Credentials2/CredentialStore.py

```diff
@@ -0,0 +1,259 @@
+
+        self.credentials.remove(credential_object)
+
+    def __str__(self, interactive=False):
+        # Get the string information with which to fill the table
+        headers = ['Type', 'Location', 'Valid', 'Time left']
+        cred_info = [[str(f) for f in (type(cred).__name__, cred.location, cred.is_valid(), cred.time_left())] for cred in self.credentials]
+        rows = [headers] + cred_info
+
+        # Get the length of the longest string in each column
+        column_widths = [
+            max(len(field) for field in column)
+            for column in zip(*rows)
+            ]
+
+        def pad_row_strings(row, widths, filler=' '):
+            # type: (List[str], List[int], str) -> List[str]
+            """Add padding to each of ``row`` to equal the corresponding ``widths`` entry, padded with ``filler``"""
+            return ['{field:{filler}<{width}}'.format(field=field[0], filler=filler, width=field[1]) for field in zip(row, widths)]
+        header_strings = pad_row_strings(headers, column_widths)
+        divider_strings = pad_row_strings([''] * len(column_widths), column_widths, filler='-')
+        row_strings = [pad_row_strings(cred, column_widths) for cred in cred_info]
+
+        # Concatenate the field string together
+        def strings_to_row(strings, spacer='|'):
+            # type: (List[str], str) -> str
+            return ' {0} '.format(spacer).join(strings)
+        header = strings_to_row(header_strings)
+        divider = strings_to_row(divider_strings, spacer='+')
+        body = '\n'.join(strings_to_row(cred) for cred in row_strings)
+
+        return '\n'.join([header, divider, body])
+
+    @export
+    def __iter__(self):
+        """Allow iterating over the store directly"""
+        # yield from self.credentialList #In Python 3.3
+        return iter(self.credentials)
+
+    @export
+    def __len__(self):
+        """How many credentials are known about in the system"""
+        return len(self.credentials)
+
+    @export
+    def __getitem__(self, query):
+        # type: (ICredentialRequirement) -> ICredentialInfo
+        """
+        This function will try quite hard to find and wrap any
+        missing credential but will never create a new file on disk.
+        It should *always* be non-interactive so that it can be run
+        on background threads without issue.
+
+        Args:
+            query (ICredentialRequirement):
+
+        Returns:
+            A single ICredentialInfo object which matches the requirements
+
+        Raises:
+            KeyError: If it could not provide a credential
+            TypeError: If query is of the wrong type
+        """
+
+        if not isinstance(query, ICredentialRequirement):
+            raise TypeError('Credential store query should be of type ICredentialRequirement')
+
+        match = self.match(query)
+        if match:
+            return match
+
+        try:
+            cred = self.create(query, create=False, check_file=True)
```

Comment:
This line is attempting to create an `ICredentialInfo` object which would try to wrap an existing credential file on disk if one exists. When Ganga starts, the credential store is empty even if there are proxy files in `/tmp`. This will try to wrap one of those files and add an entry in the credential store for it.